source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -0,0 +1,217 @@
+#include "source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.h"
+
+#include "envoy/extensions/network/dns_resolver/getaddrinfo/v3/getaddrinfo_dns_resolver.pb.h"
+#include "envoy/network/dns_resolver.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+// This resolver uses getaddrinfo() on a dedicated resolution thread. Thus, it is only suitable
+// currently for relatively low rate resolutions. In the future a thread pool could be added if
+// desired.
+class GetAddrInfoDnsResolver : public DnsResolver, public Logger::Loggable<Logger::Id::dns> {
+public:
+  GetAddrInfoDnsResolver(Event::Dispatcher& dispatcher, Api::Api& api)
+      : dispatcher_(dispatcher),
+        resolver_thread_(api.threadFactory().createThread([this] { resolveThreadRoutine(); })) {}
+
+  ~GetAddrInfoDnsResolver() override {
+    {
+      absl::MutexLock guard(&mutex_);
+      shutting_down_ = true;
+    }
+
+    resolver_thread_->join();
+  }
+
+  // DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override {
+    ENVOY_LOG(debug, "adding new query [{}] to pending queries", dns_name);
+    auto new_query = std::make_unique<PendingQuery>(dns_name, dns_lookup_family, callback);
+    absl::MutexLock guard(&mutex_);
+    pending_queries_.emplace_back(std::move(new_query));
+    return pending_queries_.back().get();
+  }
+
+  void resetNetworking() override {}
+
+private:
+  class PendingQuery : public ActiveDnsQuery {
+  public:
+    PendingQuery(const std::string& dns_name, DnsLookupFamily dns_lookup_family, ResolveCb callback)
+        : dns_name_(dns_name), dns_lookup_family_(dns_lookup_family), callback_(callback) {}
+
+    void cancel(CancelReason) override {
+      ENVOY_LOG(debug, "cancelling query [{}]", dns_name_);
+      cancelled_ = true;
+    }
+
+    const std::string dns_name_;
+    const DnsLookupFamily dns_lookup_family_;
+    ResolveCb callback_;
+    bool cancelled_{false};
+  };
+  // Must be a shared_ptr for passing around via post.
+  using PendingQuerySharedPtr = std::shared_ptr<PendingQuery>;
+
+  // Parse a GAI response and determine the final address list.
+  std::pair<ResolutionStatus, std::list<DnsResponse>>
+  processGaiResponse(const PendingQuery& query, addrinfo* addrinfo_result) {
+    std::list<DnsResponse> v4_results;
+    std::list<DnsResponse> v6_results;
+    for (auto ai = addrinfo_result; ai != nullptr; ai = ai->ai_next) {
+      if (ai->ai_family == AF_INET) {
```

Comment:
optional (given we commented on perf) but it'd be nice to skip v4/v6 inserts if they're not wanted